### ISSUE TITLE
Don't pollyfill Array.indexOf if it's already been pollyfilled

### ISSUE THREAD

**Issue by clifgriffin** [ingenesis/shopp#3292]
shopp.js polyfills `Array.indexOf` but in a way that can cause errors. It think the way Shopp is polyfilling the feature isn't 100% compatible with all uses. 

We ran into errors using when Shopp and Aesop Story Engine are loaded on the same page that has a Vimeo embed. It seems one of Aesop's scripts that deals with responsive videos is using `indexOf` in a way that triggers a security error in the browser when it tries to use Shopp's polyfill of the feature.  Essentially it's the "Can't access iframe unless domain, port, and protocol are the same" error. 

By adding a bit more protection to the polyfill declaration, I circumvented the error:

Before:
```js 
if (!Array.indexOf) { ... }
```

After:
```js 
if (!Array.indexOf && !Array.prototype.indexOf ) { ... }
```

Basically, just have Shopp stand down if the feature has already been pollyfilled. 

We might also want to look at the latest / greatest way to polyfill this feature since it's been nearly 6 years since this code was added to Shopp. 

Tagging @jonathandavis since he will need to be the one that pushes this change as he's the only one that has the rig to properly minify the scripts right now. (We should get grunt or gulp setup ;-) )

**Comment by msigley** [comment]
We should probably update the polyfill as well. An IE8 compatible polyfill is available for this at:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf

**Comment by clifgriffin** [comment]
Yeah, saw that which inspired my suggestion. Guessing the one we are using isn't quite complete based on that one.

**Comment by msigley** [comment]
Having discussed this more with @clifgriffin, I feel our search algorithm needs to be altered in the following way:

All results with a Product/Collection name match should always appear above any other result.

This wouldn't ensure that the exact match title result is at the page, but should at least ensure it is on the first page and this change should also not require a major overhaul of the search algorithm.

**Comment by clifgriffin** [comment]
@msigley Looks like you commented on the wrong issue sir.

**Comment by msigley** [comment]
Aggh your right. I deleted it and moved it to the right issue.

**Comment by jonathandavis** [comment]
There are two extra features of the other polyfill: parameter protection and support for an additional `fromIndex` parameter. Is this really necessary for a polyfill we don't really even need anymore? I'd rather drop IE8 support this cycle.

**Comment by clifgriffin** [comment]
That works for me, but we should provide a doc for people since this will effect the front end. 




It'd also be good to give people a heads up so they can plan if IE8 is still important for them. 



Clifton Griffin
Founder and Lead Developer 
CGD, Inc


Our website: http://cgd.io 
Follow me on Twitter: @clifgriffin
Awesome WordPress Plugins: http://cgd.io/store

**Comment by msigley** [comment]
I think after Windows 10 comes out the IE 8 market share on desktop will be next to nothing so dropping IE8 in the next release is a good idea.

Will IE9 be the new minimum?

-----Original Message-----

**Comment by jonathandavis** [comment]
Yes, IE9 minimum.

**Comment by clifgriffin** [comment]
Thinking about this more, if we are talking about one polyfill, I'm not sure if it is worth abandoning IE8 support for the front end store. 

Like it or not, there are still a lot of IE8 users and philosophically we may hate them but we can't exclude the fact that there are likely Shopp stores that target demographics that are disproportionately skewed to the IE8 users of the world. 

Let's just make our polyfill stand down if the feature has been added, and maybe replace it with a more compatible version if there is a better way to polyfill it now. It's super easy to fix this and doesn't risk harming any current users. 

Who is with me!??

**Comment by msigley** [comment]
@clifgriffin unfortunately IE8 support is alot more than polyfills for websites because of all of the css3 and html5 stuff added in IE9.

I agree with you though in that if supporting IE8 is as easy as adding a few polyfills, we should do it, but I think the next step for us is to look on the default shopp theme and see where IE8 is holding us back. I suppose we could look at adding an html5 stub to shopp as well.

-----Original Message-----

**Comment by clifgriffin** [comment]
@msigley We currently have IE8 support. Removing that polyfill would be removing IE8 support. So I say, it's not worth it for one polyfill.

**Comment by jonathandavis** [comment]
Microsoft officially ends all support for IE8 in January 2016. It currently holds 10% of worldwide usage. Until it drops to 1% we should still support it and as @clifgriffin said, it's an easy fix.

**Comment by clifgriffin** [comment]
@jonathandavis Can we come up with a standard way to handle the minification process with our JS files? Currently you're the only one that can update these, de facto :-)

**Comment by msigley** [comment]
@clifgriffin: I like Google's Closure Compiler service for this.
https://developers.google.com/closure/compiler/
http://www.closure-compiler.appspot.com/home

**Comment by jonathandavis** [comment]
Not true, you can download and use YUI Compressor yourself to minify the JS.

**Comment by clifgriffin** [comment]
@jonathandavis Yeah, I just would like to use the same tool / settings so that we don't end up with changesets that amount to mostly minification process differences. :-)

**Comment by jonathandavis** [comment]
We could look into commit hook processing. I'd like to do more research on best practices in that realm. Definitely needs to be automated. There are lots of tasks like this that should be automated and aren't. That seems like another project. I'll file a new issue report.

**Comment by jonathandavis** [comment]
Fixed in 74d301f
